model/publishing/delivery/tasks/RemoteTaskStatusSynchroniser.php

```diff
@@ -21,40 +28,177 @@ class RemoteTaskStatusSynchroniser implements Action,ServiceLocatorAwareInterfac
-        $taskId = array_shift($params);
-        $envId = array_shift($params);
+        list ($remoteTaskId, $envId, $deliveryUri, $testUri) = $params;
+        $this->remoteTaskId = $remoteTaskId;
+        $this->remoteEnvironmentId = $envId;
+
+        try {
+            $this->status = $this->fetchRemoteTaskStatus();
+            if (!$this->isTaskFinished($this->status)) {
+                return Report::createInfo(__('Remote task is still running.'));
+            }
 
-        $url = '/tao/TaskQueue/getStatus';
-        $request = new Request('GET', trim($url, '/').'?'.http_build_query(['id' => $taskId]));
-        $response = PlatformService::singleton()->callApi($envId, $request);
+            $remoteTaskData = $this->getRemoteTaskDetails();
+            $report = new Report(Report::TYPE_SUCCESS, __('Remote task was completed.'), $remoteTaskData);
 
-        if ($response->getStatusCode() == 200) {
-            $body = json_decode($response->getBody()->getContents(), true);
-            if ($body['success'] == true && isset($body['data'])) {
-                $this->status = $body['data'];
-                $report = new \common_report_Report(\common_report_Report::TYPE_SUCCESS, __('Status from remote successfully received.'), $this->status);
-            } else {
-                $report = new \common_report_Report(\common_report_Report::TYPE_ERROR, __('Recheck is failed'));
+            $remoteReport = Report::jsonUnserialize($remoteTaskData['report']);
+            if (!$remoteReport instanceof Report) {
+                $report->add(Report::createFailure(__('Remote task does not have a report.')));
             }
+            $report->add($this->prepareRemoteTaskExecutionReport($remoteReport));
+
+            /** @var EventManager $eventManager */
+            $remoteDeliveryId = $this->getRemoteDeliveryId($remoteReport);
+            $this->triggerRemoteDeliveryCreatedEvent($remoteDeliveryId, $deliveryUri, $testUri);
+        } catch (Exception $e) {
+            $this->getLoggerService()->logError($e->getMessage(), [$e->__toString()]);
+            $report = Report::createFailure(__('Checking remote task status failed.'));
+        }
+
+
+        return $report;
+    }
+
+    private function fetchRemoteTaskStatus(): string
+    {
+        $response = $this->callRemoteApi('/tao/TaskQueue/getStatus');
 
+        return $response['data'];
+    }
+
+    /**
+     * @return mixed
+     * @throws PublishingFailedException
+     */
+    private function getRemoteTaskDetails()
+    {
+        $response = $this->callRemoteApi('/tao/TaskQueue/get');
+
+        return $response['data'];
+    }
+
+    private function isTaskFinished(string $status): bool
+    {
+        return in_array(
+            $status,
+            [
+                CategorizedStatus::STATUS_COMPLETED,
+                CategorizedStatus::STATUS_CANCELLED,
+                CategorizedStatus::STATUS_FAILED,
+                CategorizedStatus::STATUS_ARCHIVED
+            ]
```

Comment:
The array is better off as a private constant or a static local variable IMO.
Just so that it's not redeclared upon each call.

Reply:
What do you think about:
```
private $finalTaskStatuses = [
        CategorizedStatus::STATUS_COMPLETED,
        CategorizedStatus::STATUS_CANCELLED,
        CategorizedStatus::STATUS_FAILED,
        CategorizedStatus::STATUS_ARCHIVED
    ];
```

Reply:
Why not the following?
```
private const FINAL_TASK_STATUSES = [
    CategorizedStatus::STATUS_COMPLETED,
    CategorizedStatus::STATUS_CANCELLED,
    CategorizedStatus::STATUS_FAILED,
    CategorizedStatus::STATUS_ARCHIVED
];
```

Reply:
Good as well. I missed `static` in property declaration. But probably constant will be better.

Reply:
Done